tests/bgp/test_traffic_shift.py

```diff
@@ -0,0 +1,275 @@
+            # Filter out routes announced by this neigh
+            skip = False
+            for aspath in aspaths:
+                if str(neigh_hosts[hostname]['conf']['bgp']['asn']) in aspath:
+                    skip = True
+                    break
+            if skip:
+                continue
+            if route not in routes.keys():
+                logger.warn("{} not found on {}".format(route, hostname))
+                return False
+    return True
+
+def verify_loopback_route_with_community(dut_host, neigh_hosts, ip_ver, community):
+    logger.info("Verifying only loopback routes are announced to bgp neighbors")
+    mg_facts = dut_host.minigraph_facts(host=dut_host.hostname)['ansible_facts']
+    for i in range(0, 2):
+        addr = mg_facts['minigraph_lo_interfaces'][i]['addr']
+        if ipaddress.IPNetwork(addr).version == 4:
+            lo_addr_v4 = ipaddress.IPNetwork(addr)
+        else:
+            # The IPv6 Loopback announced to neighbors is /64
+            lo_addr_v6 = ipaddress.IPNetwork(addr + "/64")
+    if 4 == ip_ver:
+        lo_addr = lo_addr_v4
+    else:
+        lo_addr = lo_addr_v6
+    routes_on_all_eos = parse_routes_on_eos(dut_host, neigh_hosts, ip_ver)
+    for hostname, routes in routes_on_all_eos.iteritems():
+        logger.info("Verifying only loopback routes(ipv{}) are announced to {}".format(ip_ver, hostname))
+        for prefix, received_community in routes.iteritems():
+            if ipaddress.IPNetwork(prefix) != lo_addr:
+                logger.warn("route for {} is found on {}, which is not in loopback address".format(prefix, hostname))
+                return False
+            if received_community != community:
+                logger.warn("community for route {} is unexpected {}".format(prefix, received_community))
+                return False
+    return True
+
+def verify_only_loopback_routes_are_announced_to_neighs(dut_host, neigh_hosts, community):
+    """
+    Verify only loopback routes with certain community are announced to neighs in TSA
+    """
+    return verify_loopback_route_with_community(dut_host, neigh_hosts, 4, community) and \
+        verify_loopback_route_with_community(dut_host, neigh_hosts, 6, community)
+
+def test_TSA(duthost, ptfhost, nbrhosts, common_setup_teardown, traffic_shift_community):
+    """
+    Test TSA
+    Verify all routes are announced to bgp monitor, and only loopback routes are announced to neighs
+    """
+    try:
+        # Issue TSA on DUT
+        duthost.shell("TSA")
+        # Verify DUT is in maintenance state.
+        pytest_assert(TS_MAINTENANCE == get_traffic_shift_state(duthost),
+                      "DUT is not in maintenance state")
+        # Start bgpmon on DUT
+        logger.info("Starting bgpmon on DUT")
+        duthost.command("sonic-cfggen -j {} -w".format(BGPMON_CONFIG_FILE))
+        time.sleep(BGP_ANNOUNCE_TIME)
+        bgpmon_routes = parse_exabgp_dump(ptfhost)
+        rib_v4 = parse_rib(duthost, 4)
```

Comment:
Should we start bgpmon before TSA?

Reply:
I think bgpmon should be started after ```TSA```. That is because if we start bgpmon, and then set ```TSA```, there will be a short ```TSB (normal)``` state between two operations, during which routes are possible to be announced. We might get a failure then.